File: unity/Rotgrid/Assets/Scripts/Weapons/WeaponSystem.cs
using System.Collections.Generic;
using UnityEngine;

namespace Rotgrid
{
    /// <summary>
    /// Owns the two carried weapons plus the knife, resolves shots against
    /// zombie hitboxes and the world, and reports hits back for scoring.
    /// </summary>
    public class WeaponSystem
    {
        const float SwapTime = 0.55f;

        readonly GameRun _run;
        readonly ViewModel _vm;
        readonly int _vmLayer;
        readonly Weapon[] _slots = new Weapon[2];
        int _index;

        float _swapT, _meleeCd, _meleeActive;
        bool _wasTriggerDown;
        public bool adsHeld;
        readonly List<float> _burstQueue = new List<float>();

        public WeaponSystem(GameRun run, ViewModel vm, int vmLayer)
        {
            _run = run;
            _vm = vm;
            _vmLayer = vmLayer;
            _slots[0] = new Weapon(Weapons.All["sidearm"]);
            _vm.SetWeapon(Current.def, _vmLayer);
        }

        public Weapon Current { get { return _slots[_index]; } }
        public Weapon Other { get { return _slots[_index == 0 ? 1 : 0]; } }
        public bool Busy { get { return _swapT > 0f || _meleeActive > 0f; } }

        public bool Ads
        {
            get { return adsHeld && !Busy && !Current.reloading && !_run.player.downed; }
        }

        float ReloadSpeedMult() { return _run.player.HasPerk("rapid") ? 2f : 1f; }
        float SwapSpeedMult() { return _run.player.HasPerk("swift") ? 1.35f : 1f; }

        // ------------------------------------------------------------ inventory
        public bool Owns(string id)
        {
            foreach (var w in _slots) if (w != null && w.Id == id) return true;
            return false;
        }

        public bool AmmoFull(string id)
        {
            foreach (var w in _slots)
                if (w != null && w.Id == id) return w.reserve >= w.def.reserve && w.mag >= w.def.mag;
            return false;
        }

        public void RefillWeapon(string id)
        {
            foreach (var w in _slots) if (w != null && w.Id == id) w.RefillAmmo();
        }

        public void RefillAll()
        {
            foreach (var w in _slots) if (w != null) w.RefillAmmo();
        }

        public Weapon GiveWeapon(WeaponDef def)
        {
            for (int i = 0; i < 2; i++)
            {
                if (_slots[i] != null && _slots[i].Id == def.id)
                {
                    _slots[i].RefillAmmo();
                    _run.Toast("AMMO REFILLED", ToastKind.Good, false);
                    return _slots[i];
                }
            }
            int target = _slots[1] == null ? 1 : _index;
            var w = new Weapon(def);
            _slots[target] = w;
            SwitchTo(target, true);
            _run.Toast(def.name, ToastKind.Good, false);
            return w;
        }

        /// <summary>Hands the active weapon to the refit machine.</summary>
        public Weapon TakeForUpgrade()
        {
            var w = Current;
            var stub = new Weapon(Weapons.All["sidearm"]);
            stub.mag = 0;
            stub.reserve = 0;
            _slots[_index] = stub;
            _vm.SetWeapon(stub.def, _vmLayer);
            return w;
        }

        public void ReturnUpgraded(Weapon weapon)
        {
            if (weapon == null) return;
            _slots[_index] = weapon;
            _vm.SetWeapon(weapon.def, _vmLayer);
            _swapT = 0.3f;
            _run.Toast(weapon.Name, ToastKind.Good, true);
        }

        void SwitchTo(int i, bool immediate)
        {
            if (i < 0 || i > 1 || _slots[i] == null) return;
            if (i == _index && !immediate) return;
            if (Current != null) Current.CancelReload();
            _index = i;
            _swapT = SwapTime / SwapSpeedMult();
            _vm.SetWeapon(_slots[i].def, _vmLayer);
        }

        public void Swap()
        {
            int next = _index == 0 ? 1 : 0;
            if (_slots[next] == null) return;
            SwitchTo(next, false);
            Game.Audio.Play(Sfx.Reload3, 0.5f);
        }

        // ------------------------------------------------------------ update
        public void Update(float dt, bool canAct)
        {
            _swapT = Mathf.Max(0f, _swapT - dt);
            _meleeCd = Mathf.Max(0f, _meleeCd - dt);
            if (_meleeActive > 0f)
            {
                _meleeActive -= dt;
                if (_meleeActive <= 0f) _vm.ShowKnife(false);
            }

            var w = Current;
            w.Update(dt, ReloadSpeedMult());

            // Queued burst shots.
            for (int i = _burstQueue.Count - 1; i >= 0; i--)
            {
                float t = _burstQueue[i] - dt;
                if (t <= 0f)
                {
                    _burstQueue.RemoveAt(i);
                    if (!Busy && !w.reloading && w.mag > 0) FireShot(w);
                }
                else _burstQueue[i] = t;
            }

            if (!canAct)
            {
                adsHeld = false;
                _wasTriggerDown = false;
                return;
            }

            adsHeld = InputMap.Down(Bind.Ads) && !_run.player.sprinting;

            if (InputMap.Pressed(Bind.Swap)) Swap();
            if (InputMap.Pressed(Bind.Slot1)) SwitchTo(0, false);
            if (InputMap.Pressed(Bind.Slot2)) SwitchTo(1, false);
            if (InputMap.Pressed(Bind.Reload)) w.StartReload(ReloadSpeedMult());
            if (InputMap.Pressed(Bind.Melee)) Melee();

            bool held = InputMap.Down(Bind.Fire);
            bool pressed = held && !_wasTriggerDown;

            if (Busy || _run.player.downed) { _wasTriggerDown = held; return; }

            // Firing a shotgun mid-reload interrupts it, like the real thing.
            if (w.reloading && w.def.shellReload && held && w.mag > 0) w.CancelReload();

            if (w.WantsToFire(held, pressed))
            {
                if (w.mag > 0)
                {
                    FireShot(w);
                    if (w.def.fireMode == FireMode.Burst) QueueBurst(w);
                }
                else if (pressed)
                {
                    Game.Audio.Play(Sfx.DryFire, 0.8f);
                    w.StartReload(ReloadSpeedMult());
                }
            }

            // Auto-reload once the magazine runs dry and the trigger is released.
            if (w.mag == 0 && !w.reloading && w.CanReload && !held) w.StartReload(ReloadSpeedMult());

            _wasTriggerDown = held;
        }

        void QueueBurst(Weapon w)
        {
            int n = w.def.burst - 1;
            for (int i = 1; i <= n; i++) _burstQueue.Add(w.FireInterval * i);
            w.burstCooldown = w.FireInterval * n + w.def.burstDelay;
        }

        // ------------------------------------------------------------ firing
        public void FireShot(Weapon w)
        {
            if (!w.Consume()) return;
            var player = _run.player;
            var def = w.def;

            player.stats.shotsFired++;
            player.stats.Weapon(w.Id).shotsFired++;

            Game.Audio.Play(SoundFor(def.sound), 1f);
            _vm.FireKick(def);
            player.AddShake(Mathf.Clamp(def.recoilV * 0.012f, 0.02f, 0.14f));

            // Recoil pattern: vertical climb plus alternating horizontal drift.
            float heat = 0.55f + w.heat * 0.9f;
            float adsFactor = Ads ? 0.62f : 1f;
            player.AddRecoil(Util.Rand(-def.recoilH, def.recoilH) * 0.006f * heat * adsFactor,
                             def.recoilV * 0.007f * heat * adsFactor);

            float spread = CurrentSpread(w);
            Vector3 origin = player.camera.position;
            Vector3 baseDir = player.camera.forward;
            Vector3 muzzle = _vm.MuzzleWorld();

            var hitThisShot = new HashSet<Zombie>();
            bool anyHit = false, anyCrit = false;
            int killed = 0;

            for (int p = 0; p < Mathf.Max(1, def.pellets); p++)
            {
                Vector3 dir = baseDir;
                if (spread > 0f)
                {
                    dir += new Vector3(Util.Rand(-spread, spread), Util.Rand(-spread, spread), Util.Rand(-spread, spread));
                    dir = dir.normalized;
                }
                bool hit, crit;
                int k;
                ResolvePellet(origin, dir, w, muzzle, hitThisShot, out hit, out crit, out k);
                anyHit |= hit;
                anyCrit |= crit;
                killed += k;
            }

            if (anyHit)
            {
                player.stats.shotsHit++;
                player.stats.Weapon(w.Id).shotsHit++;
                _run.Hitmarker(anyCrit, killed > 0);
            }
        }

        static Sfx SoundFor(SoundWeight s)
        {
            switch (s)
            {
                case SoundWeight.Light: return Sfx.ShootLight;
                case SoundWeight.Heavy: return Sfx.ShootHeavy;
                case SoundWeight.Shotgun: return Sfx.ShootShotgun;
                case SoundWeight.Energy: return Sfx.ShootEnergy;
                default: return Sfx.ShootMedium;
            }
        }

        float CurrentSpread(Weapon w)
        {
            var def = w.def;
            var player = _run.player;
            float s = Ads ? def.spreadAds : def.spreadHip;
            float moving = new Vector2(player.vel.x, player.vel.z).magnitude;
            s += def.spreadMove * Mathf.Clamp01(moving / 6f) * (Ads ? 0.35f : 1f);
            if (!player.grounded) s *= 1.6f;
            if (player.crouching) s *= 0.7f;
            s *= 1f + w.heat * 0.55f;
            return s;
        }

        void ResolvePellet(Vector3 origin, Vector3 dir, Weapon w, Vector3 muzzle,
                           HashSet<Zombie> alreadyHit, out bool anyHit, out bool crit, out int killed)
        {
            anyHit = false;
            crit = false;
            killed = 0;

            var def = w.def;
            var wallHit = _run.world.Raycast(origin, dir, def.range);
            float maxDist = wallHit.hit ? wallHit.dist : def.range;

            var hits = Physics.RaycastAll(origin, dir, maxDist);
            System.Array.Sort(hits, (a, b) => a.distance.CompareTo(b.distance));

            int budget = Mathf.Max(1, def.penetration);
            Vector3 lastPoint = Vector3.zero;
            bool hasLast = false;

            foreach (var h in hits)
            {
                if (budget <= 0) break;
                var hb = h.collider != null ? h.collider.GetComponent<ZombieHitbox>() : null;
                if (hb == null || hb.zombie == null || !hb.zombie.active) continue;
                var z = hb.zombie;
                if (alreadyHit.Contains(z)) continue;
                alreadyHit.Add(z);
                budget--;
                anyHit = true;
                lastPoint = h.point;
                hasLast = true;

                float dmg = _run.player.instantKill ? z.health * 4f + 10000f : DamageAt(def, h.distance);
                var res = z.TakeDamage(dmg, hb.part, def, _run.fx, dir);
                _run.player.stats.Weapon(w.Id).damage += Mathf.RoundToInt(res.damage);
                Game.Audio.PlayAt(res.headshot ? Sfx.HitCrit : Sfx.Hit, h.point, 0.8f);
                if (res.crit) crit = true;
                if (res.killed) { killed++; _run.OnKill(z, res.headshot, w, false, false); }
                else _run.OnHit(z, res);

                if (def.special == SpecialEffect.Chain) Chain(z, w, dmg);
                if (def.special == SpecialEffect.Burn) z.ApplyBurn(def.damage * 0.12f, 3f);
            }

            if (def.special == SpecialEffect.Blast)
            {
                Vector3 p = hasLast ? lastPoint : (wallHit.hit ? wallHit.point : origin + dir * def.range);
                Blast(p, def, w);
            }

            if (!anyHit && wallHit.hit)
            {
                _run.fx.ImpactPuff(wallHit.point, wallHit.normal);
                Game.Audio.PlayAt(Sfx.Ricochet, wallHit.point, 0.5f);
            }

            Vector3 end = hasLast ? lastPoint : (wallHit.hit ? wallHit.point : origin + dir * def.range);
            if (Random.value < 0.55f || def.pellets > 1)
                _run.fx.Tracer(muzzle, end, def.accent == 0x66d9ff ? Util.Hex(0x9fe6ff) : Util.Hex(0xffd9a0));
        }

        static float DamageAt(WeaponDef def, float dist)
        {
            float t = Mathf.Clamp01(dist / def.range);
            return def.damage * (1f - t * def.falloff);
        }

        void Chain(Zombie source, Weapon w, float baseDamage)
        {
            var def = w.def;
            var visited = new HashSet<Zombie> { source };
            var current = source;
            float dmg = baseDamage * def.chainFalloff;

            for (int i = 0; i < def.chainTargets; i++)
            {
                Zombie next = null;
                foreach (var z in _run.zombies.Near(current.pos.x, current.pos.z, def.chainRadius))
                {
                    if (visited.Contains(z)) continue;
                    next = z;
                    break;
                }
                if (next == null) break;
                visited.Add(next);
                _run.fx.Arc(current.pos + Vector3.up * 1.2f, next.pos + Vector3.up * 1.2f, Util.Hex(0x66d9ff));
                float d = _run.player.instantKill ? next.health * 4f + 10000f : dmg;
                var res = next.TakeDamage(d, BodyPart.Torso, def, _run.fx, Vector3.zero);
                if (res.killed) _run.OnKill(next, false, w, false, false);
                else _run.OnHit(next, res);
                dmg *= def.chainFalloff;
                current = next;
            }
        }

        void Blast(Vector3 point, WeaponDef def, Weapon w)
        {
            float radius = def.blastRadius;
            float damage = def.blastDamage > 0f ? def.blastDamage : def.damage;
            _run.fx.Explosion(point, radius, def.Accent);
            Game.Audio.PlayAt(Sfx.Explosion, point, 1f);
            _run.player.AddShake(0.28f);

            foreach (var z in _run.zombies.Near(point.x, point.z, radius))
            {
                float d = Mathf.Sqrt((z.pos.x - point.x) * (z.pos.x - point.x) + (z.pos.z - point.z) * (z.pos.z - point.z));
                float falloff = Mathf.Clamp01(1f - d / radius);
                float dmg = _run.player.instantKill ? z.health * 4f + 10000f : damage * falloff;
                var res = z.TakeDamage(dmg, BodyPart.Torso, def, _run.fx, Vector3.zero);
                if (res.killed) _run.OnKill(z, false, w, false, false);
                else _run.OnHit(z, res);
            }
        }

        // ------------------------------------------------------------ melee
        public void Melee()
        {
            if (_meleeCd > 0f || _run.player.downed) return;
            _meleeCd = 0.62f;
            _meleeActive = 0.42f;
            _vm.ShowKnife(true);
            _vm.StartMelee();
            Game.Audio.Play(Sfx.Melee, 0.8f);

            var cam = _run.player.camera;
            var hits = Physics.RaycastAll(cam.position, cam.forward, Weapons.Knife.range);
            System.Array.Sort(hits, (a, b) => a.distance.CompareTo(b.distance));

            Zombie target = null;
            BodyPart part = BodyPart.Torso;
            Vector3 point = cam.position + cam.forward;
            foreach (var h in hits)
            {
                var hb = h.collider != null ? h.collider.GetComponent<ZombieHitbox>() : null;
                if (hb == null || hb.zombie == null || !hb.zombie.active) continue;
                target = hb.zombie;
                part = hb.part;
                point = h.point;
                break;
            }
            if (target == null)
            {
                // Slight forgiveness: sweep for anything very close in front.
                var near = _run.zombies.Near(_run.player.pos.x + cam.forward.x * 1.2f,
                                             _run.player.pos.z + cam.forward.z * 1.2f, 1.5f);
                if (near.Count > 0) { target = near[0]; point = target.pos + Vector3.up; }
            }
            if (target == null) return;

            float dmg = _run.player.instantKill
                ? target.health * 4f + 10000f
                : Weapons.Knife.damage * (1f + _run.player.PerkCount * 0.05f);
            var res = target.TakeDamage(dmg, part, Weapons.Knife, _run.fx, cam.forward);
            Game.Audio.PlayAt(Sfx.Hit, point, 0.9f);
            _run.Hitmarker(res.headshot, res.killed);
            if (res.killed) _run.OnKill(target, res.headshot, null, true, false);
            else _run.OnHit(target, res);
        }
    }
}

File: src/weapons/WeaponSystem.js
import * as THREE from 'three';
import { Weapon } from './Weapon.js';
import { WEAPONS, KNIFE } from './WeaponDefs.js';
import { voiceFor } from './WeaponVoices.js';
import { Input } from '../core/Input.js';
import { Audio } from '../core/AudioEngine.js';
import { clamp, rand } from '../core/Util.js';

const SWAP_TIME = 0.55;

/**
 * Owns the two carried weapons plus the knife, resolves shots against zombie
 * hitboxes and the world, and reports hits back to the game for scoring.
 */
export class WeaponSystem {
  constructor(ctx) {
    this.ctx = ctx;                 // { player, world, zombies, fx, viewModel, camera, hooks }
    this.slots = [new Weapon(WEAPONS.sidearm), null];
    this.index = 0;
    this.swapT = 0;
    this.pendingIndex = null;
    this.meleeCd = 0;
    this.meleeActive = 0;
    this.raycaster = new THREE.Raycaster();
    this.raycaster.far = 220;
    this.wasTriggerDown = false;
    this.adsHeld = false;
    this._tmpDir = new THREE.Vector3();
    this._tmpOrigin = new THREE.Vector3();
    this.ctx.viewModel.setWeapon(this.current.def);
  }

  get current() { return this.slots[this.index]; }
  get other() { return this.slots[this.index === 0 ? 1 : 0]; }
  get busy() { return this.swapT > 0 || this.meleeActive > 0; }
  get ads() {
    return this.adsHeld && !this.busy && !this.current.reloading && !this.ctx.player.downed;
  }

  reloadSpeedMult() { return this.ctx.player.perks.has('rapid') ? 2.0 : 1; }
  swapSpeedMult() { return this.ctx.player.perks.has('swift') ? 1.35 : 1; }

  // ------------------------------------------------------------- inventory
  ownsWeapon(id) { return this.slots.some((w) => w && w.id === id); }

  weaponAmmoFull(id) {
    const w = this.slots.find((s) => s && s.id === id);
    return w ? w.reserve >= w.def.reserve && w.mag >= w.def.mag : false;
  }

  refillWeapon(id) {
    const w = this.slots.find((s) => s && s.id === id);
    if (w) w.refillAmmo();
  }

  refillAll() {
    for (const w of this.slots) if (w) w.refillAmmo();
  }

  giveWeapon(def) {
    const existing = this.slots.findIndex((w) => w && w.id === def.id);
    if (existing >= 0) {
      this.slots[existing].refillAmmo();
      this.ctx.hooks.toast('AMMO REFILLED', 'pts');
      return this.slots[existing];
    }
    const empty = this.slots.indexOf(null);
    const target = empty >= 0 ? empty : this.index;
    const w = new Weapon(def);
    this.slots[target] = w;
    this._switchTo(target, true);
    this.ctx.hooks.toast(def.name, 'pts');
    return w;
  }

  /** Hands the active weapon over to the refit machine. */
  takeWeaponForUpgrade() {
    const w = this.current;
    this.slots[this.index] = new Weapon(WEAPONS.sidearm);
    this.slots[this.index].mag = 0;
    this.slots[this.index].reserve = 0;
    this.ctx.viewModel.setWeapon(this.current.def);
    return w;
  }

  returnUpgraded(weapon) {
    if (!weapon) return;
    this.slots[this.index] = weapon;
    this.ctx.viewModel.setWeapon(weapon.def);
    this.swapT = 0.3;
    this.ctx.hooks.toast(weapon.name, 'pts', true);
  }

  _switchTo(i, immediate = false) {
    if (i === this.index && !immediate) return;
    if (!this.slots[i]) return;
    this.current?.cancelReload();
    this.index = i;
    this.swapT = SWAP_TIME / this.swapSpeedMult();
    this.ctx.viewModel.setWeapon(this.slots[i].def);
  }

  swap() {
    const next = this.index === 0 ? 1 : 0;
    if (!this.slots[next]) return;
    this._switchTo(next);
    Audio.play('reload', { stage: 2, volume: 0.5 });
  }

  // ------------------------------------------------------------- update
  update(dt, canAct) {
    this.swapT = Math.max(0, this.swapT - dt);
    this.meleeCd = Math.max(0, this.meleeCd - dt);
    if (this.meleeActive > 0) {
      this.meleeActive -= dt;
      if (this.meleeActive <= 0) this.ctx.viewModel.showKnife(false);
    }

    const w = this.current;
    const wasReloading = w.reloading;
    w.update(dt, this.reloadSpeedMult());
    if (wasReloading && !w.reloading) this.ctx.viewModel.endReload();

    if (!canAct) {
      this.adsHeld = false;
      this.wasTriggerDown = false;
      return;
    }

    this.adsHeld = Input.isDown('ads') && !this.ctx.player.sprinting;

    if (Input.wasPressed('swap')) this.swap();
    if (Input.wasPressed('slot1')) this._switchTo(0);
    if (Input.wasPressed('slot2')) this._switchTo(1);
    if (Input.wasPressed('reload')) {
      if (w.startReload(this.reloadSpeedMult())) this.ctx.viewModel.startReload();
    }
    if (Input.wasPressed('melee')) this.melee();

    const held = Input.isDown('fire');
    const pressed = held && !this.wasTriggerDown;
    if (!held) w.shotsThisTrigger = 0;

    if (this.busy || this.ctx.player.downed) {
      this.wasTriggerDown = held;
      return;
    }

    // Firing a shotgun mid-reload interrupts it, like the real thing.
    if (w.reloading && w.def.shellReload && held && w.mag > 0) {
      w.cancelReload();
      this.ctx.viewModel.endReload();
    }

    if (w.wantsToFire(held, pressed)) {
      if (w.mag > 0) {
        this._fireShot(w);
        if (w.def.fireMode === 'burst') this._queueBurst(w);
      } else if (pressed) {
        Audio.play('dryfire', { volume: 0.8 });
        if (w.canReload && w.startReload(this.reloadSpeedMult())) this.ctx.viewModel.startReload();
      }
    }

    // Auto-reload when the magazine runs dry mid-burst.
    if (w.mag === 0 && !w.reloading && w.canReload && !held) {
      if (w.startReload(this.reloadSpeedMult())) this.ctx.viewModel.startReload();
    }

    this.wasTriggerDown = held;
  }

  _queueBurst(w) {
    const n = (w.def.burst ?? 3) - 1;
    for (let i = 1; i <= n; i++) {
      setTimeout(() => {
        if (this.current !== w || w.mag <= 0 || w.reloading || this.busy) return;
        this._fireShot(w);
      }, i * w.fireInterval * 1000);
    }
    w.burstCooldown = w.fireInterval * n + (w.def.burstDelay ?? 0.2);
  }

  // ------------------------------------------------------------- firing
  _fireShot(w) {
    if (!w.consume()) return;
    const { player, viewModel, fx, hooks } = this.ctx;

    player.stats.shotsFired++;
    hooks.weaponStat(w.id).shotsFired++;

    const def = w.def;
    Audio.play('shoot', { weight: def.sound, voice: voiceFor(def), volume: 1 });
    viewModel.fireKick(def);
    player.addShake(clamp(def.recoil.v * 0.012, 0.02, 0.14));

    // Recoil pattern: vertical climb plus alternating horizontal drift.
    const heat = 0.55 + w.heat * 0.9;
    const adsFactor = this.ads ? 0.62 : 1;
    player.addRecoil(
      rand(-def.recoil.h, def.recoil.h) * 0.006 * heat * adsFactor,
      def.recoil.v * 0.007 * heat * adsFactor,
    );

    const spread = this._currentSpread(w);
    const origin = this.ctx.camera.getWorldPosition(this._tmpOrigin).clone();
    const baseDir = this.ctx.camera.getWorldDirection(this._tmpDir).clone();
    const muzzle = viewModel.muzzleWorld(this.ctx.camera).clone();

    const pellets = def.pellets ?? 1;
    const hitZombies = new Set();
    let anyHit = false;
    let anyCrit = false;
    let killed = 0;

    for (let p = 0; p < pellets; p++) {
      const dir = baseDir.clone();
      if (spread > 0) {
        dir.x += rand(-spread, spread);
        dir.y += rand(-spread, spread);
        dir.z += rand(-spread, spread);
        dir.normalize();
      }
      const r = this._resolvePellet(origin, dir, w, muzzle, hitZombies);
      anyHit = anyHit || r.hit;
      anyCrit = anyCrit || r.crit;
      killed += r.killed;
    }

    if (anyHit) {
      player.stats.shotsHit++;
      hooks.weaponStat(w.id).shotsHit++;
      hooks.hitmarker(anyCrit, killed > 0);
    }
  }

  _currentSpread(w) {
    const def = w.def;
    const player = this.ctx.player;
    let s = this.ads ? def.spread.ads : def.spread.hip;
    const moving = Math.hypot(player.vel.x, player.vel.z);
    s += (def.spread.move ?? 0.03) * clamp(moving / 6, 0, 1) * (this.ads ? 0.35 : 1);
    if (!player.grounded) s *= 1.6;
    if (player.crouching) s *= 0.7;
    s *= 1 + w.heat * 0.55;
    return s;
  }

  _resolvePellet(origin, dir, w, muzzle, hitZombies) {
    const { world, zombies, fx, player, hooks } = this.ctx;
    const def = w.def;
    const wallHit = world.raycastWorld(origin, dir, def.range);
    const wallDist = wallHit ? wallHit.dist : def.range;

    this.raycaster.set(origin, dir);
    this.raycaster.far = Math.min(def.range, wallDist);
    const targets = zombies.raycastTargets();
    const hits = targets.length ? this.raycaster.intersectObjects(targets, false) : [];

    let budget = def.penetration ?? 1;
    let hitAny = false;
    let crit = false;
    let killed = 0;
    let lastPoint = null;

    for (const h of hits) {
      if (budget <= 0) break;
      const z = h.object.userData.zombie;
      if (!z || !z.active) continue;
      if (hitZombies.has(z)) continue;
      hitZombies.add(z);
      budget--;
      hitAny = true;
      lastPoint = h.point;

      const dmg = player.instantKill ? z.health * 4 + 10000 : this._damageAt(def, h.distance);
      const res = z.takeDamage(dmg, h.object.userData.part, def, fx, dir);
      hooks.weaponStat(w.id).damage += Math.round(res.damage);
      Audio.playAt('hit', h.point, { crit: res.headshot, volume: 0.8 });
      if (res.crit) crit = true;
      if (res.killed) {
        killed++;
        hooks.onKill(z, { headshot: res.headshot, weapon: w, melee: false });
      } else {
        hooks.onHit(z, res);
      }

      if (def.special === 'chain') this._chain(z, w, dmg);
      if (def.special === 'burn') z.applyBurn(def.damage * 0.12, 3);
    }

    if (def.special === 'blast') {
      const p = lastPoint ?? (wallHit ? wallHit.point : {
        x: origin.x + dir.x * def.range, y: origin.y + dir.y * def.range, z: origin.z + dir.z * def.range,
      });
      this._blast(p, def, w);
    }

    if (!hitAny && wallHit) {
      fx.impactPuff(wallHit.point, wallHit.normal);
      Audio.playAt('ricochet', wallHit.point, { volume: 0.5 });
    }

    const end = lastPoint ?? (wallHit ? wallHit.point : {
      x: origin.x + dir.x * def.range, y: origin.y + dir.y * def.range, z: origin.z + dir.z * def.range,
    });
    if (Math.random() < 0.55 || def.pellets > 1) {
      fx.tracer(muzzle, end, def.accent === 0x66d9ff ? 0x9fe6ff : 0xffd9a0);
    }

    return { hit: hitAny, crit, killed };
  }

  _damageAt(def, dist) {
    const t = clamp(dist / def.range, 0, 1);
    return def.damage * (1 - t * (def.falloff ?? 0.5));
  }

  _chain(source, w, baseDamage) {
    const { zombies, fx, hooks, player } = this.ctx;
    const cfg = w.def.chain ?? { targets: 4, radius: 6, falloff: 0.7 };
    let current = source;
    let dmg = baseDamage * cfg.falloff;
    const visited = new Set([source]);
    for (let i = 0; i < cfg.targets; i++) {
      const near = zombies.near(current.pos.x, current.pos.z, cfg.radius)
        .filter((z) => !visited.has(z));
      if (!near.length) break;
      const next = near[0];
      visited.add(next);
      fx.arc(
        { x: current.pos.x, y: current.pos.y + 1.2, z: current.pos.z },
        { x: next.pos.x, y: next.pos.y + 1.2, z: next.pos.z },
        0x66d9ff,
      );
      const res = next.takeDamage(player.instantKill ? next.health * 4 + 10000 : dmg, 'torso', w.def, fx, null);
      if (res.killed) hooks.onKill(next, { headshot: false, weapon: w, melee: false });
      else hooks.onHit(next, res);
      dmg *= cfg.falloff;
      current = next;
    }
  }

  _blast(point, def, w) {
    const { zombies, fx, hooks, player } = this.ctx;
    const cfg = def.blast ?? { radius: 3.5, damage: def.damage };
    fx.explosion(point, cfg.radius, def.accent ?? 0xffa860);
    Audio.playAt('explosion', point, { volume: 1 });
    player.addShake(0.28);
    for (const z of zombies.near(point.x, point.z, cfg.radius)) {
      const d = Math.hypot(z.pos.x - point.x, z.pos.z - point.z);
      const falloff = clamp(1 - d / cfg.radius, 0, 1);
      const res = z.takeDamage(
        player.instantKill ? z.health * 4 + 10000 : cfg.damage * falloff,
        'torso', def, fx, null,
      );
      if (res.killed) hooks.onKill(z, { headshot: false, weapon: w, melee: false });
      else hooks.onHit(z, res);
    }
  }

  // ------------------------------------------------------------- melee
  melee() {
    if (this.meleeCd > 0 || this.ctx.player.downed) return;
    this.meleeCd = 0.62;
    this.meleeActive = 0.42;
    this.ctx.viewModel.showKnife(true);
    this.ctx.viewModel.startMelee();
    Audio.play('melee', { volume: 0.8 });

    const { camera, zombies, fx, hooks, player } = this.ctx;
    const origin = camera.getWorldPosition(new THREE.Vector3());
    const dir = camera.getWorldDirection(new THREE.Vector3());
    this.raycaster.set(origin, dir);
    this.raycaster.far = KNIFE.range;
    const hits = this.raycaster.intersectObjects(zombies.raycastTargets(), false);
    // Slight forgiveness: also sweep for anything very close in front.
    let target = hits.find((h) => h.object.userData.zombie?.active);
    if (!target) {
      const near = zombies.near(
        player.pos.x + dir.x * 1.2, player.pos.z + dir.z * 1.2, 1.5,
      );
      if (near.length) target = { object: { userData: { zombie: near[0], part: 'torso' } }, point: near[0].pos };
    }
    if (!target) return;

    const z = target.object.userData.zombie;
    const dmg = player.instantKill ? z.health * 4 + 10000 : KNIFE.damage * (1 + player.perks.size * 0.05);
    const res = z.takeDamage(dmg, target.object.userData.part ?? 'torso', KNIFE, fx, dir);
    Audio.playAt('hit', target.point, { volume: 0.9 });
    hooks.hitmarker(res.headshot, res.killed);
    if (res.killed) hooks.onKill(z, { headshot: res.headshot, weapon: null, melee: true });
    else hooks.onHit(z, res);
  }

  // ------------------------------------------------------------- HUD
  hudState() {
    const w = this.current;
    return {
      name: w.name,
      mag: w.mag,
      reserve: w.reserve,
      reloading: w.reloading,
      reloadProgress: w.reloadProgress,
      upgradeLevel: w.upgradeLevel,
      other: this.other ? this.other.name : null,
      ads: this.ads,
    };
  }

  reset() {
    this.slots = [new Weapon(WEAPONS.sidearm), null];
    this.index = 0;
    this.swapT = 0;
    this.ctx.viewModel.setWeapon(this.current.def);
  }
}
